tezos/timing/src/lib.rs

```diff
@@ -0,0 +1,870 @@
+                ":find_time": action_stats.find,
+                ":find_tree_time": action_stats.find_tree,
+                ":remove_time": action_stats.remove,
+            })?;
+        }
+
+        Ok(())
+    }
+
+    // Compute stats for the current block and global ones
+    fn sync_global_stats(
+        &mut self,
+        commit_time_irmin: f64,
+        commit_time_tezedge: f64,
+    ) -> Result<(), SQLError> {
+        let block_id = self.current_block.as_ref().map(|(id, _)| id.as_str());
+
+        let mut query = self.sql.prepare_cached(
+            "
+        UPDATE
+          blocks
+        SET
+          actions_count = :actions_count,
+          checkout_time_irmin = :checkout_time_irmin,
+          checkout_time_tezedge = :checkout_time_tezedge,
+          commit_time_irmin = :commit_time_irmin,
+          commit_time_tezedge = :commit_time_tezedge
+        WHERE
+          id = :block_id;
+            ",
+        )?;
+
+        query.execute(named_params! {
+            ":actions_count": &self.nactions,
+            ":checkout_time_irmin": &self.checkout_time.as_ref().map(|(irmin, _)| irmin),
+            ":checkout_time_tezedge": &self.checkout_time.as_ref().map(|(_, tezedge)| tezedge),
+            ":commit_time_irmin": &commit_time_irmin,
+            ":commit_time_tezedge": &commit_time_tezedge,
+            ":block_id": block_id
+        })?;
+
+        for action in self.global_stats.values_mut() {
+            action.compute_mean();
+        }
+        self.checkout_stats.compute_mean();
+        self.commit_stats.compute_mean();
+
+        for (root, action_stats) in self.global_stats.iter() {
+            let root = root.as_str();
+
+            self.insert_action_stats(root, "mem", &action_stats.mem)?;
+            self.insert_action_stats(root, "find", &action_stats.find)?;
+            self.insert_action_stats(root, "find_tree", &action_stats.find_tree)?;
+            self.insert_action_stats(root, "add", &action_stats.add)?;
+            self.insert_action_stats(root, "add_tree", &action_stats.add_tree)?;
+            self.insert_action_stats(root, "remove", &action_stats.remove)?;
+        }
+
+        self.insert_action_stats("commit", "commit", &self.commit_stats)?;
+        self.insert_action_stats("checkout", "checkout", &self.checkout_stats)?;
+
+        Ok(())
+    }
+
+    fn insert_action_stats(
+        &self,
+        root: &str,
+        action_name: &str,
+        range_stats: &RangeStats,
+    ) -> Result<(), SQLError> {
+        let mut query = self.sql.prepare_cached(
+            "
+        INSERT OR IGNORE INTO global_action_stats
+          (root, action_name)
+        VALUES
+          (:root, :action_name)
+            ",
+        )?;
+
+        query.execute(named_params! {
+            ":root": root,
+            ":action_name": action_name,
+        })?;
+
+        let mut query = self.sql.prepare_cached(
+            "
+        UPDATE
```

Comment:
My idea was to keep the calculation on SQlite and make them faster by adding a few indexes (what made them slow was that a full linear scan needed to be made, but indexes fix that). But now that you have written all the calculations in Rust I think it is worth it to keep it that way, because if at some point we switch from SQlite to something else, all that code will keep working.